velox/dwio/dwrf/reader/ReaderBase.cpp

```diff
@@ -154,29 +172,53 @@ ReaderBase::ReaderBase(
     input_->load(LogType::FOOTER);
   }
 
-  auto footerStream = input_->read(
-      fileLength_ - psLength_ - footerSize - 1, footerSize, LogType::FOOTER);
+  BufferPtr remainingFooterBuffer;
+  char* footerStart;
+  if (offset >= footerSize) {
+    offset -= footerSize;
+    footerStart = buf + offset;
+  } else {
+    remainingFooterBuffer =
+        AlignedBuffer::allocate<char>(footerSize, &options_.memoryPool());
+    footerStart = remainingFooterBuffer->asMutable<char>();
+    auto remaining = footerSize - offset;
+    input_
+        ->read(
+            fileLength_ - footerSize - psLength_ - 1,
+            remaining,
+            LogType::FOOTER)
+        ->readFully(footerStart, remaining);
+    memcpy(footerStart + remaining, buf, offset);
+    offset = 0;
+  }
+  auto decompressed = createDecompressedStream(
+      std::make_unique<dwio::common::SeekableArrayInputStream>(
+          footerStart, footerSize),
+      "File Footer");
   if (fileFormat() == FileFormat::DWRF) {
-    auto footer =
-        google::protobuf::Arena::CreateMessage<proto::Footer>(arena_.get());
-    ProtoUtils::readProtoInto<proto::Footer>(
-        createDecompressedStream(std::move(footerStream), "File Footer"),
-        footer);
-    footer_ = std::make_unique<FooterWrapper>(footer);
+    footer_ = parseFooter<proto::Footer>(decompressed.get(), arena_.get());
   } else {
-    auto footer = google::protobuf::Arena::CreateMessage<proto::orc::Footer>(
-        arena_.get());
-    ProtoUtils::readProtoInto<proto::orc::Footer>(
-        createDecompressedStream(std::move(footerStream), "File Footer"),
-        footer);
-    footer_ = std::make_unique<FooterWrapper>(footer);
+    footer_ = parseFooter<proto::orc::Footer>(decompressed.get(), arena_.get());
   }
+  footerBufferSize_ = offset;
 
   schema_ = std::dynamic_pointer_cast<const RowType>(
       convertType(*footer_, 0, options_.fileColumnNamesReadAsLowerCase()));
   VELOX_CHECK_NOT_NULL(schema_, "invalid schema");
 
-  // load stripe index/footer cache
+  // initialize file decrypter
+  handler_ =
+      DecryptionHandler::create(*footer_, options_.decrypterFactory().get());
+}
+
+void ReaderBase::loadCache() {
+  if (!footerBuffer_) {
```

Comment:
in which case this is null? on the second call? Can you comment? Thanks!

Reply:
Yes the second call.  Will not happen on Hive connector but showing up in some DWIO tests.